### DISCUSSION TITLE
Code completion unavailable for some dunder methods

### DISCUSSION THREAD

**Discussion by valorien** [microsoft/pylance-release#3074] [Enhancement]
<h3> Environment data </h3>

-   Language Server version: 2022.7.30
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda):  3.10
-   python.analysis.indexing: null
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

```python
class A:

    def __seti   # <- Code completion prompt doesn't offer the __setitem__ method

    def __len    # <- Code completion prompt doesn't offer the __len__ method
```

![setitem](https://user-images.githubusercontent.com/3483790/179829411-66d19993-cfbb-4a53-a4b4-4ae16107d6d0.jpg)

![len](https://user-images.githubusercontent.com/3483790/179829438-81236998-58cb-4338-9078-2f3807db4161.jpg)

**Comment by erictraut** [comment]
Methods are suggested only if they are overriding a method of the same name in a base class. In the example you provided, `A` derives from `object`, so only methods implemented in `object` are offered as completion suggestions. Other methods (dundered or otherwise) probably do not apply to the class, so they are not offered as suggestions.

**Comment by valorien** [comment]
Hi Eric, thank you very much for the prompt response and for the explanation.

I understand what you mean, however, adding dunder methods to a newly created class shouldn't just be seen in the context of its base class.

When developing a new class, it is extremely useful to have all available dunder methods offered in the completion menu.
This is true regardless of the methods defined by the parent class, and **especially** if that base class is `object` which is just a forced default and not some class explicitly and willfully subclassed by the developer.

In my opinion, `__len__`,  `__getitem__`, and most other dunder methods are the basic building blocks of classes in Python and it would be wise to have them available as code completion suggestions.

Here's the same code in PyCharm, BTW: 

![charm](https://user-images.githubusercontent.com/3483790/179854010-3de9c159-3018-4f56-a3fe-d9e319ed0e00.jpg)

Seems very convenient and useful to me.


-- Thanks again for your help. Much appreciated. --

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
